pageserver/src/tenant_guard.rs

```diff
@@ -0,0 +1,122 @@
+use crate::{
+    task_mgr::{self, spawn, TaskKind, BACKGROUND_RUNTIME},
+    tenant::{Tenant, TenantState},
+};
+use anyhow::Context;
+use std::{
+    fs,
+    ops::Deref,
+    sync::{Arc, Weak},
+};
+use tracing::{debug, error};
+
+struct TenantGuardInner {
+    tenant: Option<Tenant>,
+}
+
+impl Drop for TenantGuardInner {
+    fn drop(&mut self) {
+        let tenant = self.tenant.take().unwrap();
+        let conf = tenant.conf;
+        let tenant_id = tenant.tenant_id();
+        let future = async move {
+            debug!("shutdown tenant {tenant_id}");
+
+            // shutdown all tenant and timeline tasks: gc, compaction, page service
+            task_mgr::shutdown_tasks(None, Some(tenant_id), None).await;
+
+            // Ok, no background tasks running anymore. Flush any remaining data in
+            // memory to disk.
+            //
+            // We assume that any incoming connections that might request pages from
+            // the tenant have already been terminated by the caller, so there
+            // should be no more activity in any of the repositories.
+            if let Err(err) = tenant.checkpoint() {
+                error!("Could not checkpoint tenant {tenant_id} during shutdown: {err:?}");
+            }
+
+            // If removal fails there will be no way to successfully retry detach,
+            // because the tenant no longer exists in the in-memory map. And it needs to be removed from it
+            // before we remove files, because it contains references to tenant
+            // which references ephemeral files which are deleted on drop. So if we keep these references,
+            // we will attempt to remove files which no longer exist. This can be fixed by having shutdown
+            // mechanism for tenant that will clean temporary data to avoid any references to ephemeral files
+            let local_tenant_directory = conf.tenant_path(&tenant_id);
+            fs::remove_dir_all(&local_tenant_directory).with_context(|| {
+                format!(
+                    "Failed to remove local tenant directory '{}'",
+                    local_tenant_directory.display()
+                )
+            })?;
```

Comment:
Adding a review comment as we discussed this. Removal of local files or the equivalent of `detach_tenant` is not wanted in the normal case, as far as I remember from the RFC.

Instead I think there should be way to wait for "having zero active tenant guards" in `tenant_detach` and only then remove the files. One way which I suggested was tokio's broadcast channel. All it would need is a way to subscribe to it, and in this `Drop`, send out a message via that channel.

If there's no message being sent through the channel (it just gets closed) -- then something needs to be done for the semantics of that ... Anyway, it might be a good way to communicate for example any failure to checkpoint which happened.

Reply:
Alternative discussed: having some mutable state to enable the detaching on drop. Could be that the channel would be extra, I just think it would be nice to await for such thing to complete by default :) Could of course be that tenant detach must await for all cleanup to happen.

Reply:
+1 And one note on detach semantics. Currently detach is expected to return ok back to the client only when everything is successfully stopped. Otherwise it gives an error so client can retry the call. Doing detach via spawning another task without waiting for completion may lead to loss of the detach operation  in case when we return ok to client and then pageserver is restarted losing the background task